### PR TITLE
Iceberg metadata big refactoring

### PR DESCRIPTION
**Pull request by divanik**
### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
A lot of new features will require better code incapsulation (what relates to Iceberg metadata) and better code abstractions.  

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

#### CI Settings (Only check the boxes if you know what you are doing):
- [ ]  Allow: All Required Checks
- [ ]  Allow: Stateless tests
- [ ]  Allow: Stateful tests
- [ ]  Allow: Integration Tests
- [ ]  Allow: Performance tests
- [ ]  Allow: All Builds
- [ ]  Allow: batch 1, 2 for multi-batch jobs
- [ ]  Allow: batch 3, 4, 5, 6 for multi-batch jobs
---
- [ ]  Exclude: Style check
- [ ]  Exclude: Fast test
- [ ]  Exclude: All with ASAN
- [ ]  Exclude: All with TSAN, MSAN, UBSAN, Coverage
- [ ]  Exclude: All with aarch64, release, debug
---
- [ ]  Run only fuzzers related jobs (libFuzzer fuzzers, AST fuzzers, etc.)
- [ ]  Exclude: AST fuzzers
---
- [ ]  Do not test
- [ ]  Woolen Wolfdog
- [ ]  Upload binaries for special builds
- [ ]  Disable merge-commit
- [ ]  Disable CI cache

### CODE REVIEW

**Comment by robot-ch-test-poll2** [comment]
*This is an automated comment for commit c28ae71b3fa067c310a32f8b04707dd50f1288e7 with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/72941/c28ae71b3fa067c310a32f8b04707dd50f1288e7/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Docs check</td><td>Builds and tests the documentation</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/71117fbbf322caa635d75ff5709ea475c2ce2ec6/docs_check.html">✅ success</a></td></tr>
<tbody>
</table>
</details>

File: src/Storages/ObjectStorage/DataLakes/Iceberg/IcebergMetadata.h [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+
+    mutable std::optional<Strings> cached_files_for_current_snapshot;
+
+    ManifestList initializeManifestList(const String & manifest_list_file) const;
+
+    IcebergSnapshot getSnapshot(const String & manifest_list_file) const;
+
+    std::optional<Int32> getSchemaVersionByFileIfOutdated(String data_path) const
+    {
+        auto manifest_file_it = manifest_entry_by_data_file.find(data_path);
```
Comment (kssenii) [inline src/Storages/ObjectStorage/DataLakes/Iceberg/IcebergMetadata.h:118]: Move to cpp?

File: src/Storages/ObjectStorage/DataLakes/Iceberg/IcebergMetadata.cpp [hunk 2]
```diff
@@ -0,0 +1,429 @@
...
+    auto configuration_ptr = configuration.lock();
+    if (configuration_ptr == nullptr)
+        throw Exception(ErrorCodes::LOGICAL_ERROR, "Configuration is expired");
+
+    auto context = getContext();
+    StorageObjectStorageSource::ObjectInfo object_info(
+        std::filesystem::path(configuration_ptr->getPath()) / "metadata" / manifest_list_file);
+    auto manifest_list_buf = StorageObjectStorageSource::createReadBuffer(object_info, object_storage, context, log);
+
+    LOG_DEBUG(&Poco::Logger::get("initializeManifestList"), "Parse manifest list {}", manifest_list_file);
```
Comment (kssenii) [inline src/Storages/ObjectStorage/DataLakes/Iceberg/IcebergMetadata.cpp:326]: Save logger to class? I see it used more than once in the same method, therefore makes sense to save

File: src/Storages/ObjectStorage/DataLakes/Iceberg/IcebergMetadata.cpp [hunk 3]
```diff
@@ -0,0 +1,429 @@
...
+    auto manifest_list_file_reader
+        = std::make_unique<avro::DataFileReaderBase>(std::make_unique<AvroInputStreamReadBufferAdapter>(*manifest_list_buf));
+
+    LOG_DEBUG(&Poco::Logger::get("initializeManifestList"), "Parsed manifest list {}", manifest_list_file);
```
Comment (kssenii) [inline src/Storages/ObjectStorage/DataLakes/Iceberg/IcebergMetadata.cpp:330]: Should log level be reduced to test here?

**Review by kssenii — APPROVED** [review]

File: src/Storages/ObjectStorage/DataLakes/Iceberg/IcebergMetadata.cpp (hunk 3 above)
Comment (divanik) [reply src/Storages/ObjectStorage/DataLakes/Iceberg/IcebergMetadata.cpp:330]: Sorry, I forgot to remove debug logs, will do it all over the pull request

**Pull request merged**